sympy/functions/elementary/piecewise.py

```diff
@@ -305,13 +306,87 @@ def doit(self, **hints):
+        >>> from sympy.abc import x
+        >>> f = Piecewise((csc(1/x), x < 1), (2*x**2 + 1, x < 5), (log(x), True))
+        >>> f.is_meromorphic(x, 0)
+        False
+        >>> f.is_meromorphic(x, 2)
+        True
+        >>> f.is_meromorphic(x, 10)
+        True
+
+        """
+        if not x.is_symbol:
+            raise TypeError("{} should be of symbol type".format(x))
+        a = sympify(a)
+        upper_bound_list, lower_bound_list = [], []
+        for l, u, e, i in self._intervals(x)[1]:
+            upper_bound_list.append(u)
+            lower_bound_list.append(l)
+
+        if a not in lower_bound_list and a not in upper_bound_list:
+            for bound in self.as_expr_set_pairs():
+                if a in bound[1]:
+                    return bound[0]._eval_is_meromorphic(x, a)
+        else:
+            return None
+
     def _eval_simplify(self, **kwargs):
         return piecewise_simplify(self, **kwargs)
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
-        for e, c in self.args:
-            if c == True or c.subs(x, 0) == True:
-                return e.as_leading_term(x)
+        upper_bound_list, cond_list, lower_bound_list = [], [], []
+        for a, b, e, i in self._intervals(x)[1]:
+            upper_bound_list.append(b)
+            cond_list.append(e)
+            lower_bound_list.append(a)
+        upper = list(zip(cond_list, upper_bound_list))
+        lower = list(zip(cond_list, lower_bound_list))
+
+        # Check whether 0 is the upper boundary of a piece.
+        # If yes, we consider that piece and the next one for evaluation.
+        if S.Zero in upper_bound_list:
+            for i, pair in enumerate(upper):
+                if pair[1].is_zero:
+                    if cdir == -1:
+                        return pair[0].as_leading_term(x)
+                    else:
+                        return upper[i + 1][0].as_leading_term(x)
+
+        # Check whether 0 is the lower boundary of a piece.
+        # If yes, we consider that piece and the previous one for evaluation.
+        if S.Zero in lower_bound_list:
+            for i, pair in enumerate(lower):
+                if pair[1].is_zero:
+                    if cdir == -1:
+                        return lower[i - 1][0].as_leading_term(x)
+                    else:
+                        return pair[0].as_leading_term(x)
+
+        # Check whether 0 lies in the Interval of any piece.
+        # If yes, we consider that piece for evaluation.
+        for bound in self.as_expr_set_pairs():
+            if S.Zero in bound[1] and not S.Zero in bound[1].boundary:
+                return bound[0].as_leading_term(x)
+
+    def _at_infinity(self, x):
+        at_infinity = []
+        for i, pair in enumerate(self.as_expr_set_pairs()):
+            if pair[1].inf is S.NegativeInfinity or pair[1].sup is S.Infinity:
+                at_infinity.append(self.args[i])
+        return Piecewise(*tuple(at_infinity))
```

Comment:
There should be 2 pieces. The first condition should be `x < 0` in order to make 0 the single boundary point. The expressions should have `1/x` substituted for `x`. (After all, this is expected to be a simplified `subs(x, 1/x)`.)

Reply:
Ohh ok , I haven't disturbed the `subs(x, 1/x)` transformation from limits.py yet .
```
            if isinstance(e, Piecewise):
                e = e._at_infinity(z)
            newe = e.subs(z, 1/z)
```
So currently what happens is as follows. We eliminate all pieces in between and only keep those pieces whose lower limit is `-oo` and upper limit is `oo` . So basically `expr` changes to `expr1` everytime while we calculate limit at `oo` and `-oo` . . The subs thingy still takes place in `limits.py` . Doesn't seem wrong though , atleast I couldn't find any exceptions out of the cases i tried
```
>>> expr = Piecewise((x - 3, x <= -5), ((x + 1)**2, x < 0),(1/x, x<10), (5, True))
>>> expr._at_infinity(x)
Piecewise((x - 3, x <= -5), (5, True))
>>> limit(expr, x, oo)
5
>>> limit(expr, x, -oo)
-oo
>>> expr1 = expr._at_infinity(x)
>>> limit(expr1, x, oo)
5
>>> limit(expr1, x, -oo)
-oo
```

Reply:
What will happen to a condition of the type `x < a` if `subs(x. 1/x)` is applied?

Reply:
The test I've added would be a good example for this 
```
>>> expr2 = Piecewise((x - 3, x < 0), ((x + 1)**2, x < 10), (5, True))
>>> expr3 = expr2._at_infinity(x)
>>> expr3
Piecewise((x - 3, x < 0), (5, True))
>>> expr3.subs(x, 1/x)
Piecewise((-3 + 1/x, 1/x < 0), (5, True))
>>> newe = expr3.subs(x, 1/x)
>>> newe.leadterm(x, cdir = 1)
(5, 0)
>>> newe.leadterm(x, cdir = -1)
(1, -1)
```

Reply:
We want the condition `x < 0` for the first piece. `Piecewise((x - 3, x < 0), ((x + 1)**2, x < 10), (5, True))` is not a good example as there is `x < 0`. We should be prepared for `x < a` for any `a`.

Reply:
I'm sure it would work for any a though ! This is what we want right ? cause the logic works correctly like this .
```
>>> expr = Piecewise((x - 3, x < -2), ((x + 1)**2, x < 10), (5, True))
>>> expr._at_infinity(x)
Piecewise((x - 3, x < -2), (5, True))
>>> expr = Piecewise((x - 3, x < 2), ((x + 1)**2, x < 10), (5, True))
>>> expr._at_infinity(x)
Piecewise((x - 3, x < 2), (5, True)) 
```
I hope I am understanding what your trying to convey . is this there anything wrong with the logic above ? ( only keeping pieces with either lower bound at `-oo `or those with upper bound at `oo` fllowed by `subs(x, 1/x)` in `limits.py` . Maybe you could give me an example of what you would like , cause I'm a bit blinded now as I don't see any case which the previous commit is not able to answer !

Reply:
The final result should be a piecewise expression where 0 is a boundary point. That does not happen if `subs(x, 1/x)` is applied to Piecewise:
```
n [1]: e = Piecewise((x - 3, x < 2), (5, True))                                

In [2]: e.subs(x, 1/x)                                                          
Out[2]: 
⎧     1      1    
⎪-3 + ─  for ─ < 2  # this should be x < 0
⎨     x      x    
⎪                 
⎩  5     otherwise
```

Reply:
Cool I'll maybe do this , though I still have my doubts regarding why do you think the code I pushed latest would fail to give expected results .I'll try again maybe to explain throughly I guess ? 

1) We get rid of all pieces which won't contribute to limit at `oo/-oo` using `_at_infinity` helper method. Hence after this our expression ` e = Piecewise((x - 3, x < -2), ((x + 1)**2, x < 10), (5, True))` would become `e = Piecewise((x - 3, x < 2), (5, True)) ` considering we eliminate the piece `((x + 1)**2, x < 10)` as it won't contribute anything in this case !

2) Now we only have two pieces and subs would never disappoint us in such cases .
`newe = e.subs(x, 1/x) `, where `e` is the updated `e` . Hence we have `newe  = Piecewise((-3 + 1/x, 1/x < 2), (5, True))` and as my code uses `._intervals(x)` , we have 
```
>>> (Piecewise((-3 + 1/x, 1/x < 2), (5, True)))._intervals(x)[1]
[(1/2, oo, -3 + 1/x, 0), (-oo, 0, -3 + 1/x, 0), (-oo, oo, 5, 1)]
```
This is correctly representing the intervals we want which is `-3 + 1/x` for `Union(Interval(1/2, oo), Interval(-oo, 0))`n and 5 for the rest which would eventually be `Interval(0, 1/2)` 
![image](https://user-images.githubusercontent.com/87052487/144708794-393dd320-536d-4816-8b42-1bc887c162fd.png)

Hence from positive side (cdir = 1, which is basically the case for oo as cdir = -1, is reversed in `limits.py` by cdir = -cdir) we have the 5 case . This basically means `limit(5, x, 0, dir = '+') == limit(e, x, oo, dir = '-')`
And from the negative side (cdir = -1) we have the `-3 + 1/x` . This implies `limit(-3 + 1/x ,x, 0 , dir = '-') == limit(e, x, -oo, dir = '+') ` . 
And as we would always have only two pieces whenever we deal with limit at `oo` and `-oo` , we would always end up with a similar codeflow ,with one piece transformed on one side of `0` and another piece transformed ( `1/x` in place of `x`) on another side of `0`. I really don't see any case failing with this , maybe you could point something out if you have your doubts regarding certain expressions !

Reply:
> There should be 2 pieces. The first condition should be `x < 0` in order to make 0 the single boundary point. The expressions should have `1/x` substituted for `x`. (After all, this is expected to be a simplified `subs(x, 1/x)`.)

Resolving conversation as my latest commit implements this !